actionpack/lib/action_dispatch/middleware/flash.rb

```diff
@@ -284,11 +287,32 @@ def now_is_loaded?
         end
 
       private
+        SUPPORTED_PRIMITIVE_TYPES = [String, Array, Hash].freeze
+
+        def logger
+          @logger ||= ActiveSupport::Logger.new(STDOUT)
+        end
+
         def stringify_array(array) # :doc:
           array.map do |item|
             item.kind_of?(Symbol) ? item.to_s : item
           end
         end
+
+        def supported_primitive_type?(v)
+          v.nil? || SUPPORTED_PRIMITIVE_TYPES.any? { |type| v.is_a?(type) }
```

Comment:
Please switch from iteration over an array and using any? To using a hash with the class names as keys. It will be faster.

Reply:
@schneems I have benchmarked this and I don't see a big benefit in this case, maybe because the array has only 3 positions. If there isn't a big different I would prefer to stick to the Array version as it seems clearer to me.

Here you have the benchmark:
```ruby
require "Benchmark"
include Benchmark

SUPPORTED_PRIMITIVE_ARRAY = [String, Array, Hash].freeze
SUPPORTED_PRIMITIVE_HASH  = { String => nil, Array => nil, Hash => nil }

["String", [1, 2, 3], { a: 1, b: 2}].each do |value|
  puts "\n****************************"
  puts "Benchmarking value: #{value}"
  puts

  bmbm(6) do |test|
    test.report("Array: ") do
      SUPPORTED_PRIMITIVE_ARRAY.any? { |type| value.is_a?(type) }
    end

    test.report("Hash: ") do
      SUPPORTED_PRIMITIVE_HASH.any? { |k, _| value.is_a?(k) }
    end

    test.report("OR: ") do
      value.is_a?(String) || value.is_a?(Array) || value.is_a?(Hash)
    end
  end
end
```
And the results:
```bash
2.5.1 ~/ $ ruby bench_array_hash_or.rb

****************************
Benchmarking value: String

Rehearsal -------------------------------------------
Array:    0.000010   0.000006   0.000016 (  0.000006)
Hash:     0.000007   0.000002   0.000009 (  0.000009)
OR:       0.000003   0.000001   0.000004 (  0.000004)
---------------------------------- total: 0.000029sec

              user     system      total        real
Array:    0.000008   0.000002   0.000010 (  0.000005)
Hash:     0.000007   0.000002   0.000009 (  0.000006)
OR:       0.000005   0.000002   0.000007 (  0.000023)

****************************
Benchmarking value: [1, 2, 3]

Rehearsal -------------------------------------------
Array:    0.000005   0.000001   0.000006 (  0.000004)
Hash:     0.000007   0.000004   0.000011 (  0.000018)
OR:       0.000003   0.000001   0.000004 (  0.000003)
---------------------------------- total: 0.000021sec

              user     system      total        real
Array:    0.000024   0.000002   0.000026 (  0.000006)
Hash:     0.000008   0.000001   0.000009 (  0.000005)
OR:       0.000005   0.000001   0.000006 (  0.000005)

****************************
Benchmarking value: {:a=>1, :b=>2}

Rehearsal -------------------------------------------
Array:    0.000004   0.000001   0.000005 (  0.000005)
Hash:     0.000004   0.000001   0.000005 (  0.000004)
OR:       0.000004   0.000001   0.000005 (  0.000005)
---------------------------------- total: 0.000015sec

              user     system      total        real
Array:    0.000004   0.000001   0.000005 (  0.000004)
Hash:     0.000004   0.000001   0.000005 (  0.000004)
OR:       0.000004   0.000001   0.000005 (  0.000003)
```

Reply:
You’re still iterating through the hash by calling any? ans checking each key. Instead use it as a hash:

    SUPPORTED_PRIMITIVE_HASH[value.class]


For the detection to work you would also need the values to be truthy.

I recommend using the benchmark-ips gem over Benchmark.bmbm.

Reply:
OK, I got it wrong, this should give us a hint, now using benchmark-ips gem ❤️ 
```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "benchmark-ips"
end

require "benchmark/ips"

SUPPORTED_PRIMITIVE_ARRAY = [String, Array, Hash].freeze
SUPPORTED_PRIMITIVE_HASH  = { String => true, Array => true, Hash => true }

["String", [1, 2, 3], { a: 1, b: 2}].each do |value|
  puts "\n****************************"
  puts "Benchmarking value: #{value}"
  puts

  Benchmark.ips do |test|
    test.report("Array: ") do
      SUPPORTED_PRIMITIVE_ARRAY.any? { |type| value.is_a?(type) }
    end

    test.report("Hash: ") do
      SUPPORTED_PRIMITIVE_HASH[value.class]
    end

    test.report("OR: ") do
      value.is_a?(String) || value.is_a?(Array) || value.is_a?(Hash)
    end

    test.compare!
  end
end
```
And the results:
```bash
2.5.1 ~ $ ruby bench_array_hash_or.rb
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using benchmark-ips 2.7.2
Using bundler 1.16.1

****************************
Benchmarking value: String

Warming up --------------------------------------
             Array:    272.789k i/100ms
              Hash:    192.142k i/100ms
                OR:    336.287k i/100ms
Calculating -------------------------------------
             Array:       6.625M (± 4.6%) i/s -     33.280M in   5.035105s
              Hash:       3.442M (± 4.2%) i/s -     17.293M in   5.033748s
                OR:      10.453M (± 7.6%) i/s -     52.124M in   5.024647s

Comparison:
                OR: : 10453063.7 i/s
             Array: :  6624951.4 i/s - 1.58x  slower
              Hash: :  3441698.7 i/s - 3.04x  slower


****************************
Benchmarking value: [1, 2, 3]

Warming up --------------------------------------
             Array:    214.251k i/100ms
              Hash:    183.673k i/100ms
                OR:    295.872k i/100ms
Calculating -------------------------------------
             Array:       4.921M (± 4.7%) i/s -     24.639M in   5.018891s
              Hash:       3.444M (± 3.8%) i/s -     17.265M in   5.021349s
                OR:       8.834M (± 3.8%) i/s -     44.381M in   5.031975s

Comparison:
                OR: :  8833775.5 i/s
             Array: :  4920635.0 i/s - 1.80x  slower
              Hash: :  3443628.5 i/s - 2.57x  slower


****************************
Benchmarking value: {:a=>1, :b=>2}

Warming up --------------------------------------
             Array:    205.368k i/100ms
              Hash:    193.242k i/100ms
                OR:    287.559k i/100ms
Calculating -------------------------------------
             Array:       3.712M (± 4.3%) i/s -     18.688M in   5.044509s
              Hash:       3.443M (± 4.8%) i/s -     17.199M in   5.008349s
                OR:       7.544M (± 4.0%) i/s -     37.670M in   5.001988s

Comparison:
                OR: :  7544189.8 i/s
             Array: :  3711993.7 i/s - 2.03x  slower
              Hash: :  3442707.2 i/s - 2.19x  slower
```
After launching it several times I always get similar results. As there are only 3 supported classes, I think I will go for the OR which always performs better.

Reply:
My responses went to the wrong place here are the results with compare by identity

```
Benchmarking value: String

Warming up --------------------------------------
             Array:    268.224k i/100ms
              Hash:    318.214k i/100ms
                OR:    335.285k i/100ms
Calculating -------------------------------------
             Array:       6.201M (± 2.3%) i/s -     31.114M in   5.020387s
              Hash:       9.245M (± 3.6%) i/s -     46.459M in   5.032277s
                OR:       9.342M (± 5.6%) i/s -     46.605M in   5.004818s

Comparison:
                OR: :  9342494.7 i/s
              Hash: :  9245341.0 i/s - same-ish: difference falls within error
             Array: :  6200932.3 i/s - 1.51x  slower


****************************
Benchmarking value: [1, 2, 3]

Warming up --------------------------------------
             Array:    231.245k i/100ms
              Hash:    316.924k i/100ms
                OR:    302.787k i/100ms
Calculating -------------------------------------
             Array:       4.549M (± 2.6%) i/s -     22.893M in   5.036167s
              Hash:       9.033M (± 4.1%) i/s -     45.320M in   5.026575s
                OR:       7.770M (± 3.1%) i/s -     39.060M in   5.031966s

Comparison:
              Hash: :  9033053.8 i/s
                OR: :  7770054.8 i/s - 1.16x  slower
             Array: :  4548977.5 i/s - 1.99x  slower


****************************
Benchmarking value: {:a=>1, :b=>2}

Warming up --------------------------------------
             Array:    198.171k i/100ms
              Hash:    318.380k i/100ms
                OR:    284.343k i/100ms
Calculating -------------------------------------
             Array:       3.425M (± 2.5%) i/s -     17.241M in   5.036619s
              Hash:       9.075M (± 4.5%) i/s -     45.528M in   5.029719s
                OR:       6.503M (± 3.1%) i/s -     32.699M in   5.033228s

Comparison:
              Hash: :  9075060.8 i/s
                OR: :  6503066.2 i/s - 1.40x  slower
             Array: :  3425284.8 i/s - 2.65x  slower
```

The  only difference is adding this line:

```
SUPPORTED_PRIMITIVE_HASH.compare_by_identity.rehash
```

Hash is either the fastest or equal to or in all the cases. This will only improve if we add another supported class. Which, if we're going to support floats and numeric types I think we will.